File: scripts/lib/fetch-house-vote.ts
const CONGRESS_BASE = 'https://api.congress.gov/v3'

export interface HouseVoteMember {
  bioguide_id: string
  name: string
  party: string
  state: string
  position: string  // "Yea" | "Nay" | "Not Voting" | "Present"
}

export interface HouseVoteSummary {
  question: string
  result: string
  yea_total: number
  nay_total: number
  present_total: number
  not_voting_total: number
  yea_democrat: number
  nay_democrat: number
  yea_republican: number
  nay_republican: number
  yea_independent: number
  nay_independent: number
  members: HouseVoteMember[]
}

// Maps raw Congress.gov voteCast values to normalised labels
function normalisePosition(raw: string): string {
  const r = raw.toLowerCase()
  if (r === 'yea' || r === 'aye') return 'Yea'
  if (r === 'nay' || r === 'no')  return 'Nay'
  if (r === 'present')            return 'Present'
  return 'Not Voting'
}

// API structure: /house-vote/{congress}/{session}/{rollNumber}
// Members:       /house-vote/{congress}/{session}/{rollNumber}/members
// Response keys: houseRollCallVote (detail), houseRollCallVoteMemberVotes.results (members)
// Member fields: bioguideID, voteCast, voteParty, voteState, firstName, lastName

export async function fetchHouseVote(
  congress: number,
  rollNumber: number,
  apiKey: string,
  session: number = 1
): Promise<HouseVoteSummary | null> {
  const base = `${CONGRESS_BASE}/house-vote/${congress}/${session}/${rollNumber}`

  const res = await fetch(`${base}?format=json&api_key=${apiKey}`)
  if (!res.ok) return null
  const data = await res.json()

  const vote = data.houseRollCallVote
  if (!vote) return null

  const membersRes = await fetch(`${base}/members?format=json&limit=500&api_key=${apiKey}`)
  if (!membersRes.ok) return null
  const membersData = await membersRes.json()

  // Congress.gov returns houseRollCallVoteMemberVotes as an object with a `results` array
  const rawMembers: any[] = membersData.houseRollCallVoteMemberVotes?.results ?? []

  const members: HouseVoteMember[] = rawMembers
    .filter((m: any) => m.bioguideID)
    .map((m: any) => ({
      bioguide_id: m.bioguideID,
      name: `${m.firstName ?? ''} ${m.lastName ?? ''}`.trim(),
      party: m.voteParty ?? '',
      state: m.voteState ?? '',
      position: normalisePosition(m.voteCast ?? ''),
    }))

  // Compute aggregates from member data
  let yea_total = 0, nay_total = 0, present_total = 0, not_voting_total = 0
  let yea_democrat = 0, nay_democrat = 0
  let yea_republican = 0, nay_republican = 0
  let yea_independent = 0, nay_independent = 0

  for (const m of members) {
    const party = m.party.toUpperCase()
    const isDem = party === 'D'
    const isRep = party === 'R'

    switch (m.position) {
      case 'Yea':
        yea_total++
        if (isDem) yea_democrat++
        else if (isRep) yea_republican++
        else yea_independent++
        break
      case 'Nay':
        nay_total++
        if (isDem) nay_democrat++
        else if (isRep) nay_republican++
        else nay_independent++
        break
      case 'Present':
        present_total++
        break
      default:
        not_voting_total++
    }
  }

  return {
    question: vote.voteQuestion ?? vote.question ?? '',
    result: vote.result ?? '',
    yea_total,
    nay_total,
    present_total,
    not_voting_total,
    yea_democrat,
    nay_democrat,
    yea_republican,
    nay_republican,
    yea_independent,
    nay_independent,
    members,
  }
}

File: pipeline/lib/fetch_house_vote.py
"""
Congress.gov House Roll Call API client.
Port of scripts/lib/fetch-house-vote.ts
"""
import httpx
from dataclasses import dataclass, field
from pipeline.lib.config import CONGRESS_BASE, TIMEOUT_SHORT, PAGE_SIZE_HOUSE_MEMBERS


@dataclass
class HouseVoteMember:
    bioguide_id: str
    name: str
    party: str
    state: str
    position: str  # "Yea" | "Nay" | "Not Voting" | "Present"


@dataclass
class HouseVoteSummary:
    question: str
    result: str
    yea_total: int
    nay_total: int
    present_total: int
    not_voting_total: int
    yea_democrat: int
    nay_democrat: int
    yea_republican: int
    nay_republican: int
    yea_independent: int
    nay_independent: int
    members: list[HouseVoteMember] = field(default_factory=list)


def _normalise_position(raw: str) -> str:
    r = raw.lower()
    if r in ("yea", "aye"):
        return "Yea"
    if r in ("nay", "no"):
        return "Nay"
    if r == "present":
        return "Present"
    return "Not Voting"


def fetch_house_vote(
    congress: int,
    roll_number: int,
    api_key: str,
    session: int = 1,
) -> HouseVoteSummary | None:
    base = f"{CONGRESS_BASE}/house-vote/{congress}/{session}/{roll_number}"

    try:
        resp = httpx.get(f"{base}?format=json&api_key={api_key}", timeout=TIMEOUT_SHORT)
        if not resp.is_success:
            return None
        data = resp.json()
    except Exception:
        return None

    vote = data.get("houseRollCallVote")
    if not vote:
        return None

    try:
        members_resp = httpx.get(
            f"{base}/members?format=json&limit={PAGE_SIZE_HOUSE_MEMBERS}&api_key={api_key}", timeout=TIMEOUT_SHORT
        )
        if not members_resp.is_success:
            return None
        members_data = members_resp.json()
    except Exception:
        return None

    raw_members: list[dict] = members_data.get("houseRollCallVoteMemberVotes", {}).get("results", [])

    members: list[HouseVoteMember] = []
    for m in raw_members:
        if not m.get("bioguideID"):
            continue
        members.append(HouseVoteMember(
            bioguide_id=m["bioguideID"],
            name=f"{m.get('firstName', '')} {m.get('lastName', '')}".strip(),
            party=m.get("voteParty", ""),
            state=m.get("voteState", ""),
            position=_normalise_position(m.get("voteCast", "")),
        ))

    yea_total = nay_total = present_total = not_voting_total = 0
    yea_democrat = nay_democrat = 0
    yea_republican = nay_republican = 0
    yea_independent = nay_independent = 0

    for m in members:
        party = m.party.upper()
        is_dem = party == "D"
        is_rep = party == "R"

        if m.position == "Yea":
            yea_total += 1
            if is_dem:
                yea_democrat += 1
            elif is_rep:
                yea_republican += 1
            else:
                yea_independent += 1
        elif m.position == "Nay":
            nay_total += 1
            if is_dem:
                nay_democrat += 1
            elif is_rep:
                nay_republican += 1
            else:
                nay_independent += 1
        elif m.position == "Present":
            present_total += 1
        else:
            not_voting_total += 1

    return HouseVoteSummary(
        question=vote.get("voteQuestion") or vote.get("question") or "",
        result=vote.get("result", ""),
        yea_total=yea_total,
        nay_total=nay_total,
        present_total=present_total,
        not_voting_total=not_voting_total,
        yea_democrat=yea_democrat,
        nay_democrat=nay_democrat,
        yea_republican=yea_republican,
        nay_republican=nay_republican,
        yea_independent=yea_independent,
        nay_independent=nay_independent,
        members=members,
    )
